### PR TITLE
Implementing VecOf(p4) function

### PR DESCRIPTION
**Pull request by rmadar**
Addressing https://github.com/go-hep/hep/issues/831

### CODE REVIEW

File: fmom/ops_test.go [hunk 1]
```diff
@@ -44,6 +44,10 @@ func deepEqual(p1, p2 P4) bool {
 	return Equal(p1, p2)
 }
 
+func deepEqualP3(p1, p2 r3.Vec) bool {
+	return p1.X == p2.X && p1.Y == p2.Y && p1.Z == p2.Z
```
Comment (sbinet) [inline fmom/ops_test.go:48]:
```suggestion
	return p1 == p2
```
should just work.

and thus, `deepEqualP3` doesn't carry too much of its weight.
I'd just drop it.

File: fmom/ops_test.go [hunk 2]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
+			p:    newPxPyPzE(NewPxPyPzE(10, 0, 10, 20)),
+			want: r3.Vec{X: 10, Y: 0, Z: 10},
+		},
+
+		{
+			p:    newPxPyPzE(NewPxPyPzE(10, 10, 0, 20)),
+			want: r3.Vec{X: 10, Y: 10, Z: 0},
+		},
+	} {
+		p := tc.p.Clone()
```
Comment (sbinet) [inline fmom/ops_test.go:418]: that's not necessary, I'd just drop it (and use `tc.p` directly below)

File: fmom/ops_test.go [hunk 3]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
+
+		got := VecOf(p)
+
+		if !deepEqualP3(got, tc.want) {
```
Comment (sbinet) [inline fmom/ops_test.go:422]:
```suggestion
		if got != tc.want {
```

File: fmom/ops_test.go [hunk 4]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
+			t.Fatalf("got= %#v\nwant=%#v", got, tc.want)
```
Comment (sbinet) [inline fmom/ops_test.go:423]:
```suggestion
			t.Errorf("invalid spatial components for %#v: got= %#v\nwant=%#v", tc.p, got, tc.want)
```

**Review by sbinet — COMMENTED** [review]
thanks.
I do have a few minor comments+suggestions.

File: fmom/ops_test.go [hunk 5]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
 	}
 }
+
+func TestVecOf(t *testing.T) {
+	for _, tc := range []struct {
+		p    P4
+		want r3.Vec
+	}{
+		{
+			p:    newPxPyPzE(NewPxPyPzE(0, 10, 10, 20)),
```
Comment (sbinet) [inline fmom/ops_test.go:404]:
```suggestion
			p:    newPxPyPzE(NewPxPyPzE(0, 10, 20, 30)),
```

just to address any possible ambiguity.

File: fmom/ops_test.go [hunk 6]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
+			want: r3.Vec{X: 0, Y: 10, Z: 10},
+		},
+
+		{
+			p:    newPxPyPzE(NewPxPyPzE(10, 0, 10, 20)),
```
Comment (sbinet) [inline fmom/ops_test.go:409]:
```suggestion
			p:    newPxPyPzE(NewPxPyPzE(10, 0, 20, 30)),
```
ditto.

File: fmom/ops_test.go [hunk 7]
```diff
@@ -390,3 +394,33 @@ func TestBoostOf(t *testing.T) {
...
+			want: r3.Vec{X: 10, Y: 0, Z: 10},
+		},
+
+		{
+			p:    newPxPyPzE(NewPxPyPzE(10, 10, 0, 20)),
```
Comment (sbinet) [inline fmom/ops_test.go:414]:
```suggestion
			p:    newPxPyPzE(NewPxPyPzE(10, 20, 0, 30)),
```

File: fmom/ops_test.go (hunk 1 above)
Comment (rmadar) [reply fmom/ops_test.go:48]: Ha ok, nice! 

One curiosity question on this: I though there is no operator overloading in go ... how `obj1 == obj2` can work for a "user-defined" structure?

Comment (sbinet) [reply fmom/ops_test.go]: `==` and `!=` work for all values (except slices, the equality operator isn't defined for those: that's why one can't use slices as a key of a `map`).
so they do work nicely for `struct` values: it's just the field/field comparison (and recursively)

Comment (sbinet) [reply fmom/ops_test.go]: (that's not the same for other operators such as `<`, `>` or `!`)

**Comment by sbinet** [comment]
thanks.

**Pull request merged**